presto-main/src/main/java/com/facebook/presto/server/DynamicFilterService.java

```diff
@@ -0,0 +1,300 @@
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class DynamicFilterService
+{
+    private final Map<SourceDescriptor, DynamicFilterSummaryWithSenders> dynamicFilterSummaries = new HashMap<>();
+    private final Map<SourceDescriptor, SettableFuture<DynamicFilterSummary>> futures = new ConcurrentHashMap<>();
+
+    public void storeOrMergeSummary(String queryId, String source, int stageId, int taskId, int driverId, DynamicFilterSummary dynamicFilterSummary, int expectedSummariesCount)
+    {
+        DynamicFilterSummary mergedSummary;
+        synchronized (this) {
+            DynamicFilterSummaryWithSenders dynamicFilterSummaryWithSenders = dynamicFilterSummaries.get(SourceDescriptor.of(queryId, source));
+            checkState(dynamicFilterSummaryWithSenders != null, "Cannot store summary for not pre-registered task");
+
+            mergedSummary = dynamicFilterSummaryWithSenders.addSummary(DynamicFilterSummaryWithSenders.StageTaskKey.of(stageId, taskId), driverId, dynamicFilterSummary, expectedSummariesCount);
+        }
+
+        if (mergedSummary != null) {
+            futures.get(SourceDescriptor.of(queryId, source)).set(mergedSummary);
+        }
+    }
+
+    public synchronized void registerTasks(String source, Set<TaskId> taskIds)
+    {
+        if (taskIds.isEmpty()) {
+            return;
+        }
+
+        String queryId = taskIds.iterator().next().getQueryId().getId();
+        checkArgument(taskIds.stream().allMatch(taskId -> taskId.getQueryId().getId().equals(queryId)), "All tasks have to belong to the same query");
+
+        checkState(!dynamicFilterSummaries.containsKey(SourceDescriptor.of(queryId, source)), "Tasks already registered");
+        dynamicFilterSummaries.put(SourceDescriptor.of(queryId, source), new DynamicFilterSummaryWithSenders(taskIds.stream().map(DynamicFilterSummaryWithSenders.StageTaskKey::of).collect(toImmutableSet())));
+    }
+
+    public synchronized ListenableFuture<DynamicFilterSummary> getSummary(String queryId, String source)
+    {
+        SettableFuture<DynamicFilterSummary> future = futures.get(SourceDescriptor.of(queryId, source));
+        if (future == null) {
+            future = SettableFuture.create();
+            futures.put(SourceDescriptor.of(queryId, source), future);
+        }
+        DynamicFilterSummaryWithSenders dynamicFilterSummaryWithSenders = dynamicFilterSummaries.get(SourceDescriptor.of(queryId, source));
+        if (dynamicFilterSummaryWithSenders.getSummaryIfReady().isPresent()) {
```

Comment:
This throws NPE as statement above returns null ...`dynamicFilterSummaries.get(SourceDescriptor.of(queryId, source))` ... just adding null check would work ?